src/mscorlib/src/System/Enum.cs

```diff
@@ -46,115 +46,74 @@ private static ValuesAndNames GetCachedValuesAndNames(RuntimeType enumType, bool
             return entry;
         }
 
-        private static String InternalFormattedHexString(Object value)
+        private unsafe String InternalFormattedHexString()
         {
-            TypeCode typeCode = Convert.GetTypeCode(value);
-
-            switch (typeCode)
+            fixed (void* pValue = &JitHelpers.GetPinningHelper(this).m_data)
             {
-                case TypeCode.SByte :
-                    {
-                        Byte result = (byte)(sbyte)value;
-
-                        return result.ToString("X2", null);
-                    }
-
-                case TypeCode.Byte :
-                    {
-                        Byte result = (byte)value;
-
-                        return result.ToString("X2", null);
-                    }
-
-                case TypeCode.Boolean:
-                    {
-                        // direct cast from bool to byte is not allowed
-                        Byte result = Convert.ToByte((bool)value);
-
-                        return result.ToString("X2", null);
-                    }
-
-                case TypeCode.Int16:
-                    {
-                        UInt16 result = (UInt16)(Int16)value;
-
-                        return result.ToString("X4", null);
-                    }
-
-                case TypeCode.UInt16 :
-                    {
-                        UInt16 result = (UInt16)value;
-
-                        return result.ToString("X4", null);
-                    }
-
-                case TypeCode.Char:
-                    {
-                        UInt16 result = (UInt16)(Char)value;
-
-                        return result.ToString("X4", null);
-                    }
-
-                case TypeCode.UInt32:
-                    {
-                        UInt32 result = (UInt32)value;
-
-                        return result.ToString("X8", null);
-                    }
-
-                case TypeCode.Int32 :
-                    {
-                        UInt32 result = (UInt32)(int)value;
-
-                        return result.ToString("X8", null);
-                    }
-
-                case TypeCode.UInt64 :
+                unchecked
+                {
+                    switch (InternalGetCorElementType())
                     {
-                        UInt64 result = (UInt64)value;
-
-                        return result.ToString("X16", null);
+                        case CorElementType.I1:
+                            return ((byte)*(sbyte*)pValue).ToString("X2", null);
```

Comment:
It may be nice to fold to signed and unsigned cases together because of they are doing the same thing. It is fine for the unsafe low-level code like this to assume that signed and unsigned integers are stored in memory same way. There are number of similar switch statements (in VM) that make this assumption.

(It will also make the `unchecked